File: packages/client/src/utils/meditationStorage.ts
/**
 * Meditation Session Storage Utilities
 *
 * Handles localStorage persistence for meditation session state and configuration.
 */

import type { MeditationSessionState, MeditationDuration } from '@lifting/shared';
import { MEDITATION_SESSION_STALE_THRESHOLD_MS } from '@lifting/shared';

const SESSION_STATE_KEY = 'meditation-session-state';
const CONFIG_KEY = 'meditation-config';

const DEFAULT_DURATION: MeditationDuration = 10;

/**
 * Save the current session state to localStorage.
 */
export function saveMeditationState(state: MeditationSessionState): void {
  try {
    localStorage.setItem(SESSION_STATE_KEY, JSON.stringify(state));
  } catch (error) {
    console.warn('Failed to save meditation session state:', error);
  }
}

/**
 * Load the session state from localStorage.
 * Returns null if no saved state exists.
 */
export function loadMeditationState(): MeditationSessionState | null {
  try {
    const saved = localStorage.getItem(SESSION_STATE_KEY);
    if (saved === null) {
      return null;
    }
    return JSON.parse(saved) as MeditationSessionState;
  } catch (error) {
    console.warn('Failed to load meditation session state:', error);
    return null;
  }
}

/**
 * Check if a saved session state is stale (older than 1 hour).
 * A stale session should be silently discarded.
 */
export function isMeditationSessionStale(state: MeditationSessionState): boolean {
  const now = Date.now();

  // Check the relevant timestamp based on status
  const relevantTimestamp =
    state.status === 'paused' ? state.pausedAt : state.sessionStartedAt;

  if (relevantTimestamp === null) {
    // No timestamp means the session never really started
    return true;
  }

  return now - relevantTimestamp > MEDITATION_SESSION_STALE_THRESHOLD_MS;
}

/**
 * Clear the saved session state.
 */
export function clearMeditationState(): void {
  try {
    localStorage.removeItem(SESSION_STATE_KEY);
  } catch (error) {
    console.warn('Failed to clear meditation session state:', error);
  }
}

/**
 * Save the user's preferred duration to localStorage.
 */
export function saveMeditationConfig(duration: MeditationDuration): void {
  try {
    localStorage.setItem(CONFIG_KEY, JSON.stringify({ duration }));
  } catch (error) {
    console.warn('Failed to save meditation config:', error);
  }
}

/**
 * Load the user's preferred duration from localStorage.
 * Returns default (10 minutes) if no saved config exists.
 */
export function loadMeditationConfig(): MeditationDuration {
  try {
    const saved = localStorage.getItem(CONFIG_KEY);
    if (saved === null) {
      return DEFAULT_DURATION;
    }
    const parsed = JSON.parse(saved) as { duration: MeditationDuration };
    // Validate the duration is one of the allowed values
    if (parsed.duration === 5 || parsed.duration === 10 || parsed.duration === 20) {
      return parsed.duration;
    }
    return DEFAULT_DURATION;
  } catch (error) {
    console.warn('Failed to load meditation config:', error);
    return DEFAULT_DURATION;
  }
}

File: ios/BradOS/BradOS/Storage/MeditationStorage.swift
import Foundation

/// UserDefaults-based storage for meditation session state and configuration
final class MeditationStorage {
    static let shared = MeditationStorage()

    private let userDefaults: UserDefaults
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    // Keys
    private let sessionStateKey = "meditation-session-state"
    private let configKey = "meditation-config"

    init(userDefaults: UserDefaults = .standard) {
        self.userDefaults = userDefaults
    }

    // MARK: - Session State

    /// Save current meditation session state for crash recovery
    func saveMeditationState(_ state: MeditationSessionPersisted) {
        do {
            let data = try encoder.encode(state)
            userDefaults.set(data, forKey: sessionStateKey)
        } catch {
            print("Failed to save meditation state: \(error)")
        }
    }

    /// Load saved meditation session state
    func loadMeditationState() -> MeditationSessionPersisted? {
        guard let data = userDefaults.data(forKey: sessionStateKey) else {
            return nil
        }
        do {
            return try decoder.decode(MeditationSessionPersisted.self, from: data)
        } catch {
            print("Failed to load meditation state: \(error)")
            clearMeditationState()
            return nil
        }
    }

    /// Clear saved meditation session state
    func clearMeditationState() {
        userDefaults.removeObject(forKey: sessionStateKey)
    }

    /// Check if a saved session is stale (> 1 hour old)
    func isMeditationSessionStale(_ state: MeditationSessionPersisted) -> Bool {
        // If there's no start time, it's not valid
        guard let sessionStartedAt = state.sessionStartedAt else {
            return true
        }

        // Check if it's been paused too long
        if let pausedAt = state.pausedAt {
            let pauseDuration = Date().timeIntervalSince(pausedAt)
            if pauseDuration > MEDITATION_PAUSE_TIMEOUT {
                return true
            }
        }

        // Check if the overall session is too old
        let sessionAge = Date().timeIntervalSince(sessionStartedAt)
        return sessionAge > MEDITATION_STALE_THRESHOLD
    }

    /// Try to recover a valid session, returns nil if stale or none exists
    func recoverableSession() -> MeditationSessionPersisted? {
        guard let state = loadMeditationState() else {
            return nil
        }

        // Don't recover completed or idle sessions
        guard state.status == .active || state.status == .paused else {
            clearMeditationState()
            return nil
        }

        // Check if stale
        if isMeditationSessionStale(state) {
            clearMeditationState()
            return nil
        }

        return state
    }

    // MARK: - Configuration

    /// Save user's meditation preferences
    func saveMeditationConfig(_ config: MeditationConfig) {
        do {
            let data = try encoder.encode(config)
            userDefaults.set(data, forKey: configKey)
        } catch {
            print("Failed to save meditation config: \(error)")
        }
    }

    /// Load user's meditation preferences
    func loadMeditationConfig() -> MeditationConfig {
        guard let data = userDefaults.data(forKey: configKey) else {
            return .default
        }
        do {
            return try decoder.decode(MeditationConfig.self, from: data)
        } catch {
            print("Failed to load meditation config: \(error)")
            return .default
        }
    }
}
